perception/image_projection_based_fusion/src/segmentation_pointcloud_fusion/node.cpp

```diff
@@ -0,0 +1,139 @@
+
+  cv::Mat mask = in_image_ptr->image;
+  if (mask.cols == 0 || mask.rows == 0) {
+    return;
+  }
+  Eigen::Matrix4d projection;
+  projection << camera_info.p.at(0), camera_info.p.at(1), camera_info.p.at(2), camera_info.p.at(3),
+    camera_info.p.at(4), camera_info.p.at(5), camera_info.p.at(6), camera_info.p.at(7),
+    camera_info.p.at(8), camera_info.p.at(9), camera_info.p.at(10), camera_info.p.at(11), 0.0, 0.0,
+    0.0, 1.0;
+  geometry_msgs::msg::TransformStamped transform_stamped;
+  // transform pointcloud from frame id to camera optical frame id
+  {
+    const auto transform_stamped_optional = getTransformStamped(
+      tf_buffer_, input_mask.header.frame_id, input_pointcloud_msg.header.frame_id,
+      input_pointcloud_msg.header.stamp);
+    if (!transform_stamped_optional) {
+      return;
+    }
+    transform_stamped = transform_stamped_optional.value();
+  }
+
+  PointCloud2 transformed_cloud;
+  tf2::doTransform(input_pointcloud_msg, transformed_cloud, transform_stamped);
+
+  PointCloud output_cloud;
+
+  for (sensor_msgs::PointCloud2ConstIterator<float> iter_x(transformed_cloud, "x"),
+       iter_y(transformed_cloud, "y"), iter_z(transformed_cloud, "z"),
+       iter_orig_x(input_pointcloud_msg, "x"), iter_orig_y(input_pointcloud_msg, "y"),
+       iter_orig_z(input_pointcloud_msg, "z");
+       iter_x != iter_x.end();
+       ++iter_x, ++iter_y, ++iter_z, ++iter_orig_x, ++iter_orig_y, ++iter_orig_z) {
+    // skip filtering pointcloud behind the camera or too far from camera
+    if (*iter_z <= 0.0 || *iter_z > filter_distance_threshold_) {
+      output_cloud.push_back(pcl::PointXYZ(*iter_orig_x, *iter_orig_y, *iter_orig_z));
+      continue;
+    }
+
+    Eigen::Vector4d projected_point = projection * Eigen::Vector4d(*iter_x, *iter_y, *iter_z, 1.0);
+    Eigen::Vector2d normalized_projected_point = Eigen::Vector2d(
+      projected_point.x() / projected_point.z(), projected_point.y() / projected_point.z());
+
+    bool is_inside_image =
+      normalized_projected_point.x() > 0 && normalized_projected_point.x() < camera_info.width &&
+      normalized_projected_point.y() > 0 && normalized_projected_point.y() < camera_info.height;
+    if (!is_inside_image) {
+      output_cloud.push_back(pcl::PointXYZ(*iter_orig_x, *iter_orig_y, *iter_orig_z));
+      continue;
+    }
+
+    // skip filtering pointcloud where semantic id out of the defined list
+    uint8_t semantic_id = mask.at<uint8_t>(
+      static_cast<uint16_t>(normalized_projected_point.y()),
+      static_cast<uint16_t>(normalized_projected_point.x()));
+    if (static_cast<size_t>(semantic_id) >= filter_semantic_label_target_.size()) {
+      output_cloud.push_back(pcl::PointXYZ(*iter_orig_x, *iter_orig_y, *iter_orig_z));
+      continue;
+    }
+    if (!filter_semantic_label_target_.at(semantic_id)) {
+      output_cloud.push_back(pcl::PointXYZ(*iter_orig_x, *iter_orig_y, *iter_orig_z));
+    }
+  }
+
+  pcl::toROSMsg(output_cloud, output_pointcloud_msg);
```

Comment:
[imo]
Just some comments; not required to change now.
Since we don't use the PCL method, is it possible not to use PCL?
Converting ROS messages to PCL (and vice versa) incurs a significant computational cost, so we should try to avoid it.